#Exercise 3

In [1]:
!pip install -q pyomo

     |████████████████████████████████| 8.9 MB 12.5 MB/s 
     |████████████████████████████████| 49 kB 4.6 MB/s 


In [2]:
from pyomo.environ import *

In [3]:
import numpy as np

In [4]:
model = ConcreteModel()

In [5]:
N = 10
M = 4

In [6]:
obj_inhouse = np.array([35,50,58,60,44,39,45,55,35,40])
obj_ext = np.array([72,95,110,125,88,74,95,115,60,84])

In [7]:
constr_coeff_A = np.array([[1,1,1,1,1,1,1,1,1,1],[0.15,0.15,0.1,0.2,0.25,0.1,0.5,0.15,0.3,0.55],[0.8,0.75,0.75,0.6,0.55,0.55,0.4,0.35,0.3,0.3],[0.05,0.1,0.15,0.2,0.2,0.35,0.1,0.5,0.4,0.15]])

In [8]:
constr_rhs_b = np.array([450,0.35*450,0.45*450,0.2*450])

In [9]:
lower_bound = np.array([0,0,0,0,0,0,0,0,0,0])
upper_bound = np.array([12,9,16,25,4,5,10,13,0,6])

In [10]:
col_indices = np.arange(N)

In [11]:
model.x = Var(col_indices)
model.y = Var(col_indices)

In [12]:
model.constraints = ConstraintList()

In [13]:
row_indices = np.arange(M)
print(row_indices)

[0 1 2 3]


In [14]:
for i in row_indices:
  model.constraints.add(sum(constr_coeff_A[i][j]*model.x[j] + constr_coeff_A[i][j]*model.y[j] for j in col_indices)>=constr_rhs_b[i])

In [15]:
for i in col_indices:
  model.x[i].setlb(lower_bound[i])
  model.y[i].setlb(lower_bound[i])
  model.x[i].setub(upper_bound[i])

In [16]:
model.obj = Objective(expr = sum(obj_inhouse[i]*model.x[i]+obj_ext[i]*model.y[i] for i in col_indices), sense = minimize)

In [17]:
model.pprint()

3 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    4 : {1, 2, 3, 4}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   10 : {0, 1, 2, 3, 4, 5, 6, 7, 8, 9}
    y_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   10 : {0, 1, 2, 3, 4, 5, 6, 7, 8, 9}

2 Var Declarations
    x : Size=10, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  None :    12 : False :  True :  Reals
          1 :     0 :  None :     9 : False :  True :  Reals
          2 :     0 :  None :    16 : False :  True :  Reals
          3 :     0 :  None :    25 : False :  True :  Reals
          4 :     0 :  None :     4 : False :  True :  Reals
          5 :     0 :  None :     5 : False :  True :  Reals
          6 

In [18]:
!apt-get install -y -qq glpk-utils

Selecting previously unselected package libsuitesparseconfig5:amd64.
(Reading database ... 160837 files and directories currently installed.)
Preparing to unpack .../libsuitesparseconfig5_1%3a5.1.2-2_amd64.deb ...
Unpacking libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libamd2:amd64.
Preparing to unpack .../libamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libcolamd2:amd64.
Preparing to unpack .../libcolamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libcolamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libglpk40:amd64.
Preparing to unpack .../libglpk40_4.65-1_amd64.deb ...
Unpacking libglpk40:amd64 (4.65-1) ...
Selecting previously unselected package glpk-utils.
Preparing to unpack .../glpk-utils_4.65-1_amd64.deb ...
Unpacking glpk-utils (4.65-1) ...
Setting up libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Setting up libcolamd2:amd64 (1:5.1.2-2) ...
Setting up libamd2:amd64 

In [19]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: 31013.4
  Upper bound: 31013.4
  Number of objectives: 1
  Number of constraints: 5
  Number of variables: 21
  Number of nonzeros: 81
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.0081634521484375
# ----------------------------------------------------------
#   Solution Information
# --------------------------------

In [20]:
print('Objective = ', model.obj())

print('\nDecision Variables')
for i in col_indices:
  print('x[',i,'] = ', model.x[i].value)
  print('y[',i,'] = ', model.y[i].value)

print('\nConstraints')
model.constraints.display()

Objective =  31013.4

Decision Variables
x[ 0 ] =  12.0
y[ 0 ] =  79.0
x[ 1 ] =  9.0
y[ 1 ] =  0.0
x[ 2 ] =  16.0
y[ 2 ] =  0.0
x[ 3 ] =  25.0
y[ 3 ] =  0.0
x[ 4 ] =  4.0
y[ 4 ] =  0.0
x[ 5 ] =  5.0
y[ 5 ] =  0.0
x[ 6 ] =  10.0
y[ 6 ] =  0.0
x[ 7 ] =  0.0
y[ 7 ] =  0.0
x[ 8 ] =  0.0
y[ 8 ] =  120.4
x[ 9 ] =  6.0
y[ 9 ] =  163.6

Constraints
constraints : Size=4
    Key : Lower : Body  : Upper
      1 : 450.0 : 450.0 :  None
      2 : 157.5 : 157.5 :  None
      3 : 202.5 : 202.5 :  None
      4 :  90.0 :  90.0 :  None


## Ans 1 :

###Decision Variables :

$x_i = \text{amount of alloy A-i bought in house}\\y_i = \text{amount of alloy A-i bought in the market}$

###Objective Function :

$\text{min 35}x_1+50x_2+56x_3+60x_4+44x_5+39x_6+45x_7+55x_8+35x_9+40x_{10}+72y_1+95y_2+110y_3+126y_4+88y_5+74y_6+95y_7+115y_8+60y_9+84y_{10}$

###subject to:

$\bullet\sum_{i=1}^{10}x_i+y_i = 450\\
\bullet0.15(x_1+y_1) + 0.15(x_2+y_2) + 0.1(x_3+y_3) + 0.2(x_4+y_4) + 0.25(x_5+y_5) + 0.1(x_6+y_6) + 0.5(x_7+y_7) + 0.15(x_8+y_8) +0.3(x_9+y_9)+0.55(x_{10}+y_{10}) = 157.5\\
\bullet0.8(x_1+y_1) + 0.75(x_2+y_2) + 0.75(x_3+y_3) + 0.6(x_4+y_4) + 0.55(x_5+y_5) + 0.55(x_6+y_6) + 0.4(x_7+y_7) + 0.35(x_8+y_8) +0.3(x_9+y_9)+0.3(x_{10}+y_{10}) = 202.5\\
\bullet0.05(x_1+y_1) + 0.1(x_2+y_2) + 0.15(x_3+y_3) + 0.2(x_4+y_4) + 0.2(x_5+y_5) + 0.35(x_6+y_6) + 0.1(x_7+y_7) + 0.5(x_8+y_8) +0.4(x_9+y_9)+0.15(x_{10}+y_{10}) = 90\\
\bullet x_1\leq12, x_2\leq9, x_3\leq16, x_4\leq25, x_5\leq4, x_6\leq5, x_7\leq10, x_8\leq13, x_9\leq0, x_{10}\leq6\\
\bullet x_i\geq0, y_i\geq0;i = 1,\cdots,10$


##Ans 3 :

The optimal solution is 31013.4 at which the values of decision variables are as follows:

$x_1 = 12\\x_2 = 9\\x_3 = 16\\x_4 = 25\\x_5 = 4\\x_6 = 5\\x_7 = 10\\x_8 = 0\\x_9 = 0\\x_{10} = 6\\y_1 = 79\\y_2 = 0\\y_3 = 0\\y_4 = 0\\y_5 = 0\\y_6 = 0\\y_7 = 0\\y_8 = 0\\y_9 = 120.4\\y_{10} = 163.6$

The optimal value is 31013.4.

## Ans 4 :

In terms of quantity, alloy A-1, A-9 and A-10 are ranked highest among all the alloys with quantities 91, 120.4, and 169.6 respectively.

The alloys ranking the lowest are A-8, A-5 and A-6 with quantities 0, 4 and 5 respectively.

##Ans 5 :

Except A-8 and A-9, every alloy's inhouse stock is completely utilized, it happens because the cost of the alloy in the market for all these alloys, i.e., A-1, A-2, A-3, A-4, A-5, A-6, A-7 and A-10 is comparatively very high than what their inhouse cost is which is why only when the inhouse stock is completely utilized, the alloy is bought from the market (if bought).
A-9's inhouse stock is already 0.

##Ans 6 :

Yes, alloys A-2, A-3, A-4, A-5, A-6, A-7 and A-8, need not be acquired from the market since according to the composition requirement, the requirement of the alloys is fulfilled by the quantity available in the inhouse stocks.